ENH: Support fspath protocol

TomAugspurger opened the pull request:
Supports the [fspath protocol](https://www.python.org/dev/peps/pep-0519/) in most readers and writers.

closes https://github.com/pandas-dev/pandas/issues/13823
xref https://github.com/pandas-dev/pandas/pull/14123

TomAugspurger committed:
ENH: Support fspath protocol

Ensures that most of pandas readers and writers will honor the fspath
protocol, if an object defines it.

TST: remove old xfails

TomAugspurger commented:
Wait till 0.21 on this? Technically this _could_ be an API change for if people using `ExcelFile` directly, as I stringify `self.io`. I can probably work around this.

This will cause one error in the pytables tests, since *closed* HDFStore objects don't allow object access, so I can't check `hasattr(obj, '__fspath__')`. I think I'll change `HDFStore.__getattr__`, but have to run for now.

Also, any feedback on the testing method? Typically we have the tests split by the kind of reader / writer. But for things like these protocols, it's nice to test the interface against every reader / writer that should support it, so I put in in `common.py`.

jreback commented on pandas/tests/io/test_common.py:
```diff
@@ -89,6 +107,68 @@ def test_iterator(self):
         tm.assert_frame_equal(first, expected.iloc[[0]])
         tm.assert_frame_equal(concat(it), expected.iloc[1:])
 
+    @pytest.mark.parametrize('reader, module, path', [
```
if you are moving these here. then maybe do the same with Path/localpath tests (#16292). can do after of course.

though not sure whether I like better in each module, or centralized.

cc @chris-b1

TomAugspurger replied:
> though not sure whether I like better in each module, or centralized.

I like both :) Per-module for things specific to that, and centralized for "interface"-type things like this. The importskips are kind of annoying though. Happy to split this if you want.

I guess one alternative would be some kind of abstract base class for this set of tests, that would require methods to "implement" tests for `test_fspath`, `test_compression`, etc. That seems like a bit much to me though.

jreback replied:
yeah, I think this is fine actually, though #16292 would need to change (you can rebase on top maybe)?

TomAugspurger replied:
Yeah, going to do it on top #16292. We'll still need it for pathlib on 3.5, which doesn't implement __fspath__.

TomAugspurger committed:
API: Raise AttributeError on closed HDFStore

Previously, we called _check_if_open, which would raise a ClosedFileError
whenever the desired attribute wasn't found. This prevented the check required
for PEP519 to work properly, since hasattr shouldn't raise that error.

TomAugspurger commented:
https://github.com/pandas-dev/pandas/pull/16301/commits/8c10784f816ef9d7bb224df335229bdbf02a368f has the API changes required on HDFStore. Relatively minor, so I'm ok with just changing it in 0.21, but we can do a deprecation if you want, or I can just code around it in `_stringify_path` and we never change it.

jreback commented:
> pandas-dev/pandas@8c10784 has the API changes required on HDFStore. Relatively minor, so I'm ok with just changing it in 0.21, but we can do a deprecation if you want, or I can just code around it in _stringify_path and we never change it.

ok to just change this

TomAugspurger committed:
ENH: add __fspath__ to pandas own file-like objects

- HDFStore
- ExcelFile

(cherry picked from commit f9baec385e2513234a69a05031d2753899290d38)

TomAugspurger commented:
Rebased on top of #16292 

I also added a third commit adding `__fspath__` to our own path-like objects: HDFStore and ExcelFile. Do we have any others?

chris-b1 commented:
We also have `ExcelWriter`

TomAugspurger commented on pandas/tests/io/test_pytables.py:
```diff
@@ -5007,8 +5007,9 @@ def test_read_hdf_errors(self):
-                pytest.raises(NotImplementedError, read_hdf, store, 'df')
+
+    def test_read_hdf_generic_buffer_errors(self):
+        pytest.raises(NotImplementedError, read_hdf, BytesIO(b''), 'df')
```
@jreback I had to modify this test to hit the block [here](https://github.com/pandas-dev/pandas/pull/16301/files#diff-1b15d1477da3a0548d2dd72a5d023d00R321). I'm pretty sure you were just using `store` as an example of a buffer that isn't an HDFStore or filepath. But with these changes it'll be coerced to a path.

jreback replied:
sure

TomAugspurger commented:
@jreback do we have any official policy on whether or not to include package data in the distribution? This test uses a pickle that's not included, so it was failing on the build test. https://github.com/pandas-dev/pandas/pull/16301/commits/e3b3fcb9c672bab35d6d2773e6ef8de8cbd87e4b works around it.

Right now our wheel is ~38M unzipped, and about 15M of that are test files.

jreback commented:
> @jreback do we have any official policy on whether or not to include package data in the distribution? This test uses a pickle that's not included, so it was failing on the build test. e3b3fcb works around it.

we DO include things, but should only be smallish test files. Don't work around things like that.

jreback commented:
I think your file is not included, because its not in the filter for package_data (for ``.mp``)

TomAugspurger commented:
> I think your file is not included, because its not in the filter for package_data (for .mp)

Completely forgot that I added the messagepack file :/ Fixed now.

All green, other than that matplotlib test, which I'm going to debug now.

jreback commented:
@TomAugspurger lgtm. plotting occasional failures are unrelated and I think you have an issue for that anyhow.

TomAugspurger commented:
Thanks. I'm (trying to) debug the plotting stuff on my Travis branch, where it's currently not failing :/
